keps/sig-cluster-lifecycle/kubeadm/2568-kubeadm-non-root-control-plane/README.md

```diff
@@ -0,0 +1,1069 @@
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+If we hard coded the `UID` and `GID`, we could end up in a scenario where those are in use by another process on the machine, which would expose some of the credentials accessible to the `UID` and `GID`s to that process. So we plan to use adduser --system or using the appropriate ranges from /etc/login.defs instead of hard coding the `UID` and `GID`.
+
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+In `kubeadm` the control-plane components are run as [static-pods](https://kubernetes.io/docs/concepts/workloads/pods/#static-pods), i.e. pods directly managed by kubelet. We can use the `runAsUser` and `runAsGroup` fields in [SecurityContext](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#securitycontext-v1-core) to run the containers in these Pods with a unique `UID` and `GID` and the `capabilities` field to drop all capabilities other than the ones required.
+
+### Assigning UID and GID
+There are 3 options for setting the `UID`/`GID` of the control-plane components:-
+
+1. **Update the kubeadm API to make the uid/gid configurable by the user:**  This can be implemented in 2 ways:-
+    1. This can be done by adding fields `UID` and `GID` of type `int64` to the `ControlPlaneComponent` `struct` in https://github.com/kubernetes/kubernetes/blob/854c2cc79f11cfb46499454e7717a86d3214e6b0/cmd/kubeadm/app/apis/kubeadm/types.go#L132 as demonstrated in [this](https://github.com/kubernetes/kubernetes/pull/99753) PR.
+    2. User provides `kubeadm` with a range of uid/gids that are safe to use i.e. are not being used by another user.
+
+2. **Use constant values for uid/gid:** The `UID` and `GID`for each of the control-plane components would be set to some predetermined value in https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/constants/constants.go.
+
+3. **Create system users and let users override the defaults:** Use `adduser --system` or equivalent to create `UID`s in the SYS_UID_MIN - SYS_UID_MAX range and groupadd --system or equivalent to create `GID`s in the SYS_GID_MIN - SYS_GID_MAX range. Additionally if users want to specify their own `UID`s or `GID`s we will support that through `kubeadm` v1beta3 API.
+
+The author(s) believes that starting out with a safe default of option 3. and allowing the user to set the `UID` and `GID` through the API is more user friendly for users who care about which `UID`s and `GID`s that they want to run the control-plane as and also users who just wan't to quickly bootstrap. Further this feature will be opt-in and will be hidden behind a feature-gate.
```

Comment:
i still thing this would be the default option unless the user is explicit about uid/gid.

however, if we are picking random Ids one question i would have is what would be the best way for users to find them out in a deterministic way? should they just look at the component manifests or folder permissions?

Reply:
Looking at the runAsUser and runAsGroup fields in the manifest is the way to determine what the uid and gid is.